Update 7zip and SevenZipSharp

doug24 opened the pull request:
Version 18.1.0.0  (2018-01-28) and new version of SevenZipSharp from source at https://github.com/dnGrep/SevenZipSharp

JVimes commented:
Any idea what our customization of SevenZipSharp is doing? The official SevenZipSharp NuGet [package](https://www.nuget.org/packages/SevenZipSharp.Net45/) was updated 6 months ago, could we use it and remove the DLLs from our repo?

doug24 commented:
The original SevenZipSharp was abandoned on CodePlex around 2010.  There are about 5 clones of the original CodePlex code on GitHub, and many forks of those.  The NuGet package you linked looks the most up to date to me, and the one I forked for this build.  The native code 7z.dll in that repo are at least a couple of years old, but I can't tell for sure because the dlls are missing the original version data.  But that's not a big deal as we can embed a newer version of those dlls from the 7zip kits.  

I started with a fork of the https://github.com/tomap/SevenZipSharp, and ran a debug build of SevenZipSharp.dll with dnGrep.  When running with the latest 7z.dll, there were two null reference exceptions while processing different archive files.  I fixed those two exceptions.  I also added a missing file signature for .cpio archives.
